src/agent/Core/Controller/CheckoutSession.cpp

```diff
@@ -178,29 +179,52 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 	}
 }
 
+void
+Controller::handleSessionInitiationError(const SystemException &e, Request *req, Client* client, Controller* that) {
+	//TODO: only retry the session initiation if error is recoverable
+	if (req->sessionCheckoutTry < MAX_SESSION_CHECKOUT_TRY) {
+		SKC_DEBUG_FROM_STATIC(that, client, "Error initiating session (" << e.what() << "); retrying (attempt " << int(req->sessionCheckoutTry) << ")");
+		that->getContext()->libev->runLater(boost::bind(checkoutSessionLater, req));
+	} else {
+		string message = "error initiating a session (";
+		message.append(e.what());
+		message.append(")");
+		that->disconnectWithError(&client, message);
+	}
+}
+
 void
 Controller::initiateSession(Client *client, Request *req) {
 	TRACE_POINT();
 	req->sessionCheckoutTry++;
 	try {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
+			SKC_DEBUG(client, "Session socket connection in progress");
+
+			req->connectedWatcher.data = req;
+			ev_io_init(&req->connectedWatcher, onSessionSocketConnected, req->session->fd(), EV_WRITE);
+			ev_io_start(getLoop(), &req->connectedWatcher);
+			ev_timer_start(getLoop(), &req->connectedWatcherTimout);
+
+			req->connectedWatcherTimout.data = req;
+			ev_timer_init(&req->connectedWatcherTimout, onSessionSocketConnectTimeout, req->options.connectTimeout / 1000.0, 0);
```

Comment:
It's not entirely clear to me whether we're handling seconds <-> milliseconds conversion properly. constants.rb is in milliseconds but it seems web server options are not. Need to double check later when you're done removing the ApplicationPool option.